File: src/main/scala/edu/berkeley/cs/rise/opaque/OpaqueRPCListener.scala
package edu.berkeley.cs.rise.opaque

import java.util.logging.Logger

import com.google.protobuf.ByteString

import io.grpc.{Server, ServerBuilder}
import java.io.{BufferedReader, BufferedWriter, InputStreamReader, OutputStreamWriter}

import rpc.Rpc
import rpc.OpaqueRPCGrpc

import scala.concurrent.{ExecutionContext, Future}
import org.apache.spark.sql.SparkSession

import implicits._

/**
 * [[https://github.com/grpc/grpc-java/blob/v0.15.0/examples/src/main/java/io/grpc/examples/helloworld/HelloWorldServer.java]]
 */
object OpaqueRPCListener {
  private val logger = Logger.getLogger(classOf[OpaqueRPCListener].getName)

  def main(args: Array[String]): Unit = {
    val server = new OpaqueRPCListener(ExecutionContext.global)
    server.start(args)
    server.blockUntilShutdown()
    println("Print hello world server")
  }

  private val port = 50051
}

class OpaqueRPCListener(executionContext: ExecutionContext) { self =>
  private[this] var server: Server = null
  private[this] var spark: SparkSession = null
  private[this] var reader: BufferedReader = null
  private[this] var writer: BufferedWriter = null

  // TODO: Currently hard-coded, but can make server generate random string for each query
  private[this] var magicString: String = "hello world"
  private[this] var magicQuery: String = "println(\"" + magicString + "\")" + "\n"
 
  private def start(args: Array[String]): Unit = {

    // initialize process and read until empty

    val spark_home = sys.env("SPARK_HOME")
    val spark_shell_bin = spark_home + "/" + "/bin/spark-shell"
    
    // First argument should be jar file. Second argument should be spark cluster
    var opaque_jar: String = null
    var spark_master: String = null
    if (args.length == 2) {
      opaque_jar = args(0)
      spark_master = args(1)
    }

    // TODO: Make it so that the shell can run with multiple arguments
    // Create process and obtain stream input and output buffers
    val builder = new ProcessBuilder(spark_shell_bin, "--jars", opaque_jar, "--master", spark_master);
    builder.redirectErrorStream(true);
    val process = builder.start();
    val stdin = process.getOutputStream ();
    val stdout = process.getInputStream ();
    reader = new BufferedReader (new InputStreamReader(stdout));
    writer = new BufferedWriter(new OutputStreamWriter(stdin));
    
    // Import Opaque libraries and flush spark initialization
    var methodChain = 
      """
        import edu.berkeley.cs.rise.opaque.implicits._

        edu.berkeley.cs.rise.opaque.Utils.initSQLContext(spark.sqlContext)
      """
    methodChain = methodChain + magicQuery
    writer.write(methodChain)
    writer.flush()
    var line: String = ""
    while ((line = reader.readLine ()) != null && ! line.trim().equals(magicString)) {
      println(line)
    }

    // initialize SparkSession and ensure remote attestation works
    // TODO: chnage to perform RA through the shell and file system
    spark = SparkSession.builder()
      .appName("QEDBenchmark")
      .getOrCreate()
    Utils.initSQLContext(spark.sqlContext)

    // Create rpc Server and make it listen
    server = ServerBuilder.forPort(OpaqueRPCListener.port).addService(new OpaqueRPCImpl()).build.start
    OpaqueRPCListener.logger.info("Server started, listening on " + OpaqueRPCListener.port)
    sys.addShutdownHook {
      System.err.println("*** shutting down gRPC server since JVM is shutting down")
      reader.close()
      writer.close()
      self.stop()
      System.err.println("*** server shut down")
    }
  }

  private def stop(): Unit = {
    if (server != null) {
      server.shutdown()
    }
  }

  private def blockUntilShutdown(): Unit = {
    if (server != null) {
      server.awaitTermination()
    }
  }

  private class OpaqueRPCImpl extends OpaqueRPCGrpc.OpaqueRPCImplBase {

    override def relayGenerateReport(req: Rpc.RARequest, responseObserver: io.grpc.stub.StreamObserver[Rpc.RAReply]) = {
      
      val (enclave, eid) = Utils.initEnclave()
      val msg1 = enclave.GenerateReport(eid)

      val byteString: ByteString = ByteString.copyFrom(msg1)

      val reply = Rpc.RAReply.newBuilder()
	.setReport(byteString)
	.setSuccess(true)
	.build();

      responseObserver.onNext(reply);
      responseObserver.onCompleted();
    }

    override def relayQuery(req: Rpc.QueryRequest, responseObserver: io.grpc.stub.StreamObserver[Rpc.QueryReply]) = {
      var reply = Rpc.QueryReply.newBuilder().build();
      try {

        var query = req.getSqlQuery()
        query = query + "\n" + magicQuery;

        writer.write(query)
        writer.flush()

        var line: String = ""
        val replyBuilder = Rpc.QueryReply.newBuilder().setSuccess(true)
        while ((line = reader.readLine ()) != null && ! line.trim().equals(magicString)) {

          // We don't want to include duplicate lines from shell
          if (!line.startsWith("scala>")) { replyBuilder.addData(line)}
        }
        reply = replyBuilder.build();

      } catch {
        case _ : Throwable => println("Invalid Command")
        reply = Rpc.QueryReply.newBuilder()
	  .setSuccess(false)
	  .build()
      } finally {
        responseObserver.onNext(reply);
        responseObserver.onCompleted();
      }
    }

    override def relayFinishAttestation(req: Rpc.KeyRequest, responseObserver: io.grpc.stub.StreamObserver[Rpc.KeyReply]) = {
      if (req.getNonNull()) {

	// Obtain key from message and store
	val (enclave, eid) = Utils.initEnclave()
	enclave.FinishAttestation(eid, req.getKey().toByteArray())

        val reply = Rpc.KeyReply.newBuilder()
	  .setSuccess(true)
	  .build();
	responseObserver.onNext(reply);
        responseObserver.onCompleted();
      } else {
	println("Failed to receive data or client did not confirm")
	val reply = Rpc.KeyReply.newBuilder()
          .setSuccess(false) 
          .build();
	responseObserver.onNext(reply);
        responseObserver.onCompleted();
      }
    }
  }
}

File: src/main/grpc/OpaqueRPCListener.py
from concurrent import futures

import grpc

import os
import sys, getopt
import atexit
from pexpect import replwrap

import rpc_pb2
import rpc_pb2_grpc

# Remove the first line of the output which is a repeat of the input
def clean_shell_output(output):
  parsed_output = output.split("\n",1)[1]
  return parsed_output

class OpaqueRPCListener(rpc_pb2_grpc.OpaqueRPCServicer):

  def __init__(self):

    spark_home = os.getenv("SPARK_HOME")
    opaque_jar = ''
    master = ''

    # Determine opaque jar and ip of master
    argv = sys.argv[1:]
    try:
      opts, args = getopt.getopt(argv,"hj:m:",["jar=","master="])
    except getopt.GetoptError:
      print('OpaqueRPCListener.py -j <opaque jar> -m <master cluster ip>')
      sys.exit(2)
    for opt, arg in opts:
      if opt == '-h':
         print('OpaqueRPCListener.py -j <opaque jar> -m <master cluster ip>')
         sys.exit()
      elif opt in ("-j", "--jar"):
         opaque_jar = arg
      elif opt in ("-m", "--master"):
         master = arg

    # TODO: Currently hardcoding memory capacities. Will need to change to be dynamic.
    spark_shell_bin = spark_home +  "/bin/spark-shell" + " --jars " + opaque_jar + " --master" + master\
      + "--driver-memory 1G --executor-memory=2G"

    self.proc = replwrap.REPLWrapper(spark_shell_bin, "scala> ", prompt_change=None)
    print("Instantiate subprocess")

    self.proc.run_command("import edu.berkeley.cs.rise.opaque.implicits._")
    self.proc.run_command("edu.berkeley.cs.rise.opaque.Utils.initSQLContext(spark.sqlContext)")
    print("Imported Opaque libraries")

  def relayGenerateReport(self, request, context):
    output = self.proc.run_command("print_report")
    clean_output = clean_shell_output(output)
    reply = rpc_pb2.RAReply(success = True, report = clean_output)
    return reply

  ''' TODO: Do not need repeated field anymore. Can remove and only send string back.
  Other things that I can add for a better shell:
  1. Handle multi-line input
  2. Allow for up/down arrow of input (this should be done on client side)
  '''
  def relayQuery(self, request, context):

    if not request.query: 
      return rpc_pb2.QueryReply(success = True, data = "")

    output = self.proc.run_command(request.query)
    clean_output = clean_shell_output(output)
    reply = rpc_pb2.QueryReply(success = True, data = clean_output)
    return reply

  def relayFinishAttestation(self, request, context):
    print("Hello World - Finish Attestation")

def clean_up():
  print("Server shutoff")

def serve():
  server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
  rpc_pb2_grpc.add_OpaqueRPCServicer_to_server(OpaqueRPCListener(), server)
  server.add_insecure_port('[::]:50051')
  server.start()
  server.wait_for_termination()
  print("Hello world - serve")

if __name__ == '__main__':
    atexit.register(clean_up)
    serve()